tests/unit/test_entrypoints_ui_views.py

```diff
@@ -141,4 +141,32 @@ def test_create_diff_incorrect_dataset_name(mock_create_diff, ui_client):
     soup = BeautifulSoup(response.data, 'html.parser')
     body = soup.get_text()
 
-    assert 'Dataset name cannot contain spaces' in body
+    assert 'Dataset name can contain only letters, numbers, _, :, and -' in body
+
+
+@patch('rdf_differ.entrypoints.ui.views.get_report')
+def test_download_report_success(mock_get_report, ui_client):
+    dataset_id = 'dataset'
+    mock_get_report.return_value = b'important report', 200
+
+    response = ui_client.get(f'/diff-report/{dataset_id}')
+    assert 'important report' in response.data.decode()
+
+
+@patch('rdf_differ.entrypoints.ui.views.get_datasets')
+@patch('rdf_differ.entrypoints.ui.views.get_report')
+def test_download_report_failure(mock_get_report, mock_get_datasets, ui_client):
```

Comment:
are we missing an abstraction somewhere and we need to rely on patches? Can this evolve into using fake services or injecting some dependency somewhere?

Reply:
yeah, dependency injection should allow us to ditch most of the mocks